srv/src/main/java/my/bookshop/handlers/HierarchyHandler.java

```diff
@@ -0,0 +1,253 @@
+    }
+
+    private List<GenreHierarchy> topLevelsLimit(long limit, CqnPredicate filter) {
+        Map <Integer, GenreHierarchy> lookup = new HashMap<>();
+
+        CqnSelect getRoots = Select.from(AdminService_.GENRE_HIERARCHY).where(gh -> gh.parent_id().eq(0).and(filter));
+        List<GenreHierarchy> roots = db.run(getRoots).listOf(GenreHierarchy.class);
+        roots.forEach(root -> {
+            root.setDistanceFromRoot(0l);
+            lookup.put(root.getNodeId(), root);
+            List<Integer> parents = List.of(root.getNodeId());
+            for (long i = 1; i < limit; i++) {
+                List<Integer> ps = parents;
+                CqnSelect getChildren = Select.from(AdminService_.GENRE_HIERARCHY).where(gh -> gh.parent_id().in(ps).and(filter));
+                List<GenreHierarchy> children = db.run(getChildren).listOf(GenreHierarchy.class);
+                if (children.isEmpty()) {
+                    break;
+                }
+                long dfr = i;
+                parents = children.stream().peek(gh -> {
+                    gh.setParent(lookup.get(gh.getParentId()));
+                    gh.setDistanceFromRoot(dfr);
+                    lookup.put(gh.getNodeId(), gh);
+                }).map(GenreHierarchy::getNodeId).toList();
+            }
+        });
+
+        return lookup.values().stream().sorted(new Sorter()).toList();
+    }
+
+    private List<GenreHierarchy> topLevelsAll(CqnPredicate filter) {
+        Map<Integer, GenreHierarchy> lookup = new HashMap<>();
+
+        CqnSelect allCqn = Select.from(AdminService_.GENRE_HIERARCHY).where(filter);
+        var all = db.run(allCqn).listOf(GenreHierarchy.class);
+        all.forEach(gh -> lookup.put(gh.getNodeId(), gh));
+        all.forEach(gh -> gh.setParent(lookup.get(gh.getParentId())));
+        all.forEach(gh -> gh.setDistanceFromRoot(distanceFromRoot(gh)));
+
+        return all.stream().sorted(new Sorter()).toList();
+    }
+
+    private static long distanceFromRoot(GenreHierarchy gh) {
+        long dfr = 0;
+        while (gh.getParent() != null) {
+            dfr++;
+            gh = gh.getParent();
+        }
+
+        return dfr;
+    }
+
+    private class Sorter implements Comparator<GenreHierarchy> {
+
+        @Override
+        public int compare(GenreHierarchy gh1, GenreHierarchy gh2) {
+            Stack<String> path1 = getPath(gh1);
+            Stack<String> path2 = getPath(gh2);
+            int res = 0;
+
+            while (!path1.isEmpty() && !path2.isEmpty()) {
+                String last1 = path1.pop();
+                String last2 = path2.pop();
+                res = last1.compareTo(last2);
+                if (res != 0) {
+                    return res;
+                }
+            }
+            return res;
+        }
+
+        Stack<String> getPath(GenreHierarchy gh){
+            Stack<String> path = new Stack<>();
```

Comment:
I've tried a `List` here, but taking the last element doesn't look nice. `Deque` looks like an overkill for such use case.

Reply:
There is an issue with `java.util.Stack`: `java.util.Stack` extends `java.util.Vector`. But `java.util.Vector` is a very old Java class and is _thread-safe_. It contains a lot of synchronization, which makes it slow. Even worse, synchronization in the Java memory model introduces a memory-barrier for all concurrent threads, which might or might not use `Vector` or `Stack`. And a memory-barrier is something really bad: assume you have concurrent thread, which run in parallel on multiple CPU cores. They typically do their work independent of each other and use their CPU core's registers. If one thread hits a memory barrier all need to flush their registers to memory so that the synchronization can work properly. This is a quite expensive operation.

See: https://stackoverflow.com/questions/12524826/why-should-i-use-deque-over-stack

So, overall: `java.util.Vector` and `java.util.Stack` are evil.